presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java

```diff
@@ -902,7 +891,7 @@ public void addQueryInfoStateChangeListener(StateChangeListener<QueryInfo> state
 
     public ListenableFuture<QueryState> getStateChange(QueryState currentState)
     {
-        return queryState.getStateChange(currentState);
+        return Futures.transform(queryState.getStateChange(new QueryStateHolder(currentState)), QueryStateHolder::getState, directExecutor());
```

Comment:
For `currentState=FAILED` `QueryStateHolder::new` should throw, as this creates inconsistent state.
But FAILED is a terminal state, so you can handle this here:
```
if (currentState.isDone()) {
  verify(queryState.get().getState() == currentState, "Invalid currentState");
  return immediateFuture(currentState);
}
...
```